File: 234.palindrome-linked-list.cpp
/*
 * @lc app=leetcode id=234 lang=cpp
 *
 * [234] Palindrome Linked List
 */
#include <iostream>
using namespace std;
// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};
// @lc code=start
class Solution {
 public:
  bool isPalindrome(ListNode *head) {
    ListNode *fast = head;
    ListNode *slow = head;
    ListNode *reverse = nullptr;

    while (fast != nullptr && fast->next != nullptr) {
      ListNode *temp = reverse;
      ListNode *temp2 = slow;

      slow = slow->next;
      fast = fast->next->next;

      reverse = temp2;
      reverse->next = temp;
    }

    if (fast != nullptr) {
      slow = slow->next;
    }

    while (slow != nullptr) {
      if (slow->val != reverse->val) {
        return false;
      }
      slow = slow->next;
      reverse = reverse->next;
    }

    return true;
  }
};
// @lc code=end

File: 234.palindrome-linked-list.js
/*
 * @lc app=leetcode id=234 lang=javascript
 *
 * [234] Palindrome Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
  var slow = head;
  var fast = head;
  var reverseNode = null

  while (fast !== null && fast.next !== null) {
    var temp = reverseNode;
    reverseNode = new ListNode(slow.val, temp);
      
    fast = fast.next.next;
    slow = slow.next;
  }

  if (fast !== null) {
    slow = slow.next;
  }
  
  while (slow !== null) {
    if (slow.val !== reverseNode.val) {
      return false;
    }

    slow = slow.next;
    reverseNode = reverseNode.next;
  }

  return true;
};
// @lc code=end

